Segment-Store Error : SSLHandshakeException when TLS is enabled

If TLS is anabled pravega-benchmark is not working

the client is able to communicate properly with the pravega-controller but then it fails to connect on the SS (segment-store). 


> 2020-04-14 16:31:30:828 +0000 [main] INFO io.pravega.keycloak.client.PravegaKeycloakCredentials - Loaded Keycloak Credentials
2020-04-14 16:31:30:828 +0000 [main] INFO io.pravega.client.ClientConfig - Client credentials were extracted from environment variables. They weren't explicitly supplied as a Credentials object or system properties.
2020-04-14 16:31:30:829 +0000 [main] INFO io.pravega.client.stream.impl.ControllerImpl - Controller client connecting to server at pravega-controller.minikube.FQDN-OBFUSCATED:443
2020-04-14 16:31:30:930 +0000 [main] INFO io.pravega.client.admin.impl.StreamManagerImpl - Creating scope/stream: benchmark/benchmark1 with configuration: StreamConfiguration(scalingPolicy=ScalingPolicy(scaleType=FIXED_NUM_SEGMENTS, targetRate=0, scaleFactor=0, minNumSegments=1), retentionPolicy=null, timestampAggregationTimeout=0)
2020-04-14 16:31:30:956 +0000 [StreamManager-Controller-1] INFO io.pravega.client.stream.impl.ControllerImpl - [requestId=-6210969470396107479] Tagging client request (createStream-benchmark-benchmark1).
2020-04-14 16:31:31:058 +0000 [fetch-controllers-1] INFO io.pravega.client.stream.impl.ControllerResolverFactory - Attempting to refresh the controller server endpoints
2020-04-14 16:31:31:298 +0000 [fetch-controllers-1] INFO io.pravega.client.stream.impl.ControllerResolverFactory - Updating client with controllers: [[[pravega-controller.minikube.FQDN-OBFUSCATED/10.247.191.79:443]/{}]]
2020-04-14 16:31:31:431 +0000 [fetch-controllers-1] INFO io.pravega.client.stream.impl.ControllerResolverFactory - Rescheduling ControllerNameResolver task for after 120000 ms
2020-04-14 16:31:32:606 +0000 [grpc-default-executor-1] INFO io.pravega.client.stream.impl.ControllerImpl - [requestId=-6210969470396107479] Stream created successfully: benchmark1
2020-04-14 16:31:32:997 +0000 [main] INFO io.pravega.keycloak.client.PravegaKeycloakCredentials - Loaded Keycloak Credentials
2020-04-14 16:31:32:997 +0000 [main] INFO io.pravega.client.ClientConfig - Client credentials were extracted from environment variables. They weren't explicitly supplied as a Credentials object or system properties.
2020-04-14 16:31:33:035 +0000 [main] INFO io.pravega.client.stream.impl.ClientFactoryImpl - Creating writer: 0b375a0a-f9c9-4e70-9494-16f1799857fd for stream: benchmark1 with configuration: EventWriterConfig(initalBackoffMillis=1, maxBackoffMillis=20000, retryAttempts=10, backoffMultiple=10, enableConnectionPooling=true, transactionTimeoutTime=89999, automaticallyNoteTime=false)
2020-04-14 16:31:33:059 +0000 [main] INFO io.pravega.client.stream.impl.SegmentSelector - Refreshing segments for stream StreamImpl(scope=benchmark, streamName=benchmark1)
2020-04-14 16:31:33:065 +0000 [fetch-controllers-1] INFO io.pravega.client.stream.impl.ControllerResolverFactory - Attempting to refresh the controller server endpoints
2020-04-14 16:31:33:065 +0000 [fetch-controllers-1] INFO io.pravega.client.stream.impl.ControllerResolverFactory - Updating client with controllers: [[[pravega-controller.minikube.FQDN-OBFUSCATED/10.247.191.79:443]/{}]]
2020-04-14 16:31:33:068 +0000 [fetch-controllers-1] INFO io.pravega.client.stream.impl.ControllerResolverFactory - Rescheduling ControllerNameResolver task for after 120000 ms
2020-04-14 16:31:33:703 +0000 [clientInternal-2-1] INFO io.pravega.client.segment.impl.SegmentOutputStreamImpl - Fetching endpoint for segment benchmark/benchmark1/0.#epoch.0, writer 3b8f2f2d-0aba-4a4d-8722-bbf4ccbc4a3e
2020-04-14 16:31:33:707 +0000 [ForkJoinPool-2-worker-1] INFO io.pravega.perf.WriterWorker - EventsWriter: Running
2020-04-14 16:31:33:802 +0000 [clientInternal-2-1] INFO io.pravega.client.segment.impl.SegmentOutputStreamImpl - Establishing connection to PravegaNodeUri(endpoint=nautilus-pravega-segmentstore-0.pravega.minikube.FQDN-OBFUSCATED, port=12345) for benchmark/benchmark1/0.#epoch.0, writerID: 3b8f2f2d-0aba-4a4d-8722-bbf4ccbc4a3e
2020-04-14 16:31:33:812 +0000 [clientInternal-2-1] INFO io.pravega.client.netty.impl.ConnectionPoolImpl - Creating a new connection to PravegaNodeUri(endpoint=nautilus-pravega-segmentstore-0.pravega.minikube.FQDN-OBFUSCATED, port=12345)
2020-04-14 16:31:33:851 +0000 [clientInternal-2-1] INFO io.pravega.client.netty.impl.FlowHandler - Creating Flow 1 for endpoint nautilus-pravega-segmentstore-0.pravega.minikube.FQDN-OBFUSCATED. The current Channel is null.
2020-04-14 16:31:34:116 +0000 [epollEventLoopGroup-4-1] INFO io.pravega.client.netty.impl.FlowHandler - Connection established with endpoint nautilus-pravega-segmentstore-0.pravega.minikube.FQDN-OBFUSCATED on channel [id: 0x7060461a]
2020-04-14 16:31:34:276 +0000 [epollEventLoopGroup-4-1] INFO io.pravega.client.netty.impl.FlowHandler - Connection drop observed with endpoint nautilus-pravega-segmentstore-0.pravega.minikube.FQDN-OBFUSCATED
2020-04-14 16:31:34:277 +0000 [epollEventLoopGroup-4-1] WARN io.pravega.client.segment.impl.SegmentOutputStreamImpl - Failing connection for writer 3b8f2f2d-0aba-4a4d-8722-bbf4ccbc4a3e with exception io.pravega.shared.protocol.netty.ConnectionFailedException: Connection dropped for writer 3b8f2f2d-0aba-4a4d-8722-bbf4ccbc4a3e
2020-04-14 16:31:34:278 +0000 [epollEventLoopGroup-4-1] INFO io.pravega.client.segment.impl.SegmentOutputStreamImpl - Handling exception io.pravega.shared.protocol.netty.ConnectionFailedException: Connection dropped for writer 3b8f2f2d-0aba-4a4d-8722-bbf4ccbc4a3e for connection io.pravega.client.netty.impl.ClientConnectionImpl@16c228b0 on writer 3b8f2f2d-0aba-4a4d-8722-bbf4ccbc4a3e. SetupCompleted: false, Closed: false
2020-04-14 16:31:34:278 +0000 [epollEventLoopGroup-4-1] WARN io.pravega.client.segment.impl.SegmentOutputStreamImpl - Connection for segment benchmark/benchmark1/0.#epoch.0 on writer 3b8f2f2d-0aba-4a4d-8722-bbf4ccbc4a3e failed due to: Connection dropped for writer 3b8f2f2d-0aba-4a4d-8722-bbf4ccbc4a3e
2020-04-14 16:31:34:279 +0000 [epollEventLoopGroup-4-1] WARN io.pravega.client.segment.impl.SegmentOutputStreamImpl - 3b8f2f2d-0aba-4a4d-8722-bbf4ccbc4a3e Failed to connect: 
java.util.concurrent.CompletionException: io.pravega.shared.protocol.netty.ConnectionFailedException: Connection dropped for writer 3b8f2f2d-0aba-4a4d-8722-bbf4ccbc4a3e
	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:273)
	at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:280)
	at java.util.concurrent.CompletableFuture.uniExceptionally(CompletableFuture.java:888)
	at java.util.concurrent.CompletableFuture$UniExceptionally.tryFire(CompletableFuture.java:866)
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:488)
	at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1990)
	at io.pravega.client.segment.impl.SegmentOutputStreamImpl$State.failConnection(SegmentOutputStreamImpl.java:222)
	at io.pravega.client.segment.impl.SegmentOutputStreamImpl$State.access$700(SegmentOutputStreamImpl.java:99)
	at io.pravega.client.segment.impl.SegmentOutputStreamImpl.failConnection(SegmentOutputStreamImpl.java:549)
	at io.pravega.client.segment.impl.SegmentOutputStreamImpl.access$500(SegmentOutputStreamImpl.java:70)
	at io.pravega.client.segment.impl.SegmentOutputStreamImpl$ResponseProcessor.connectionDropped(SegmentOutputStreamImpl.java:306)
	at io.pravega.client.netty.impl.FlowHandler.lambda$channelUnregistered$0(FlowHandler.java:221)
	at java.util.concurrent.ConcurrentHashMap.forEach(ConcurrentHashMap.java:1597)
	at io.pravega.client.netty.impl.FlowHandler.channelUnregistered(FlowHandler.java:218)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:181)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:167)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelUnregistered(AbstractChannelHandlerContext.java:160)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelUnregistered(ChannelInboundHandlerAdapter.java:53)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:181)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:167)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelUnregistered(AbstractChannelHandlerContext.java:160)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelUnregistered(ChannelInboundHandlerAdapter.java:53)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:181)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:167)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelUnregistered(AbstractChannelHandlerContext.java:160)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelUnregistered(ChannelInboundHandlerAdapter.java:53)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:181)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:167)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelUnregistered(AbstractChannelHandlerContext.java:160)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelUnregistered(DefaultChannelPipeline.java:1412)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:181)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:167)
	at io.netty.channel.DefaultChannelPipeline.fireChannelUnregistered(DefaultChannelPipeline.java:865)
	at io.netty.channel.AbstractChannel$AbstractUnsafe$8.run(AbstractChannel.java:830)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:404)
	at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:326)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:897)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:748)
Caused by: io.pravega.shared.protocol.netty.ConnectionFailedException: Connection dropped for writer 3b8f2f2d-0aba-4a4d-8722-bbf4ccbc4a3e
	... 30 more


See error below coming from the SS
> 2020-04-14 16:33:28,166 341326926 [epollEventLoopGroup-11-2] ERROR i.p.s.p.n.ExceptionLoggingHandler - Uncaught exception on connection /172.17.0.1:61466
io.netty.handler.codec.DecoderException: javax.net.ssl.SSLHandshakeException: error:100000f7:SSL routines:OPENSSL_internal:WRONG_VERSION_NUMBER
        at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:472)
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:278)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1434)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:965)
        at io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:799)
        at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:421)
        at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:321)
        at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:897)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.lang.Thread.run(Thread.java:748)
Caused by: javax.net.ssl.SSLHandshakeException: error:100000f7:SSL routines:OPENSSL_internal:WRONG_VERSION_NUMBER
        at io.netty.handler.ssl.ReferenceCountedOpenSslEngine.sslReadErrorResult(ReferenceCountedOpenSslEngine.java:1140)
        at io.netty.handler.ssl.ReferenceCountedOpenSslEngine.unwrap(ReferenceCountedOpenSslEngine.java:1101)
        at io.netty.handler.ssl.ReferenceCountedOpenSslEngine.unwrap(ReferenceCountedOpenSslEngine.java:1169)
        at io.netty.handler.ssl.ReferenceCountedOpenSslEngine.unwrap(ReferenceCountedOpenSslEngine.java:1212)
        at io.netty.handler.ssl.SslHandler$SslEngineType$1.unwrap(SslHandler.java:216)
        at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1297)
        at io.netty.handler.ssl.SslHandler.decodeNonJdkCompatible(SslHandler.java:1211)
        at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1245)
        at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:502)
        at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:441)
        ... 14 common frames omitted
2020-04-14 16:33:28,166 341326926 [epollEventLoopGroup-11-2] ERROR i.p.s.s.h.h.ServerConnectionInboundHandler - Caught exception on connection:
io.netty.handler.codec.DecoderException: javax.net.ssl.SSLHandshakeException: error:100000f7:SSL routines:OPENSSL_internal:WRONG_VERSION_NUMBER
        at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:472)
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:278)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1434)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:965)
        at io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:799)
        at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:421)
        at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:321)
        at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:897)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.lang.Thread.run(Thread.java:748)
Caused by: javax.net.ssl.SSLHandshakeException: error:100000f7:SSL routines:OPENSSL_internal:WRONG_VERSION_NUMBER
        at io.netty.handler.ssl.ReferenceCountedOpenSslEngine.sslReadErrorResult(ReferenceCountedOpenSslEngine.java:1140)
        at io.netty.handler.ssl.ReferenceCountedOpenSslEngine.unwrap(ReferenceCountedOpenSslEngine.java:1101)
        at io.netty.handler.ssl.ReferenceCountedOpenSslEngine.unwrap(ReferenceCountedOpenSslEngine.java:1169)
        at io.netty.handler.ssl.ReferenceCountedOpenSslEngine.unwrap(ReferenceCountedOpenSslEngine.java:1212)
        at io.netty.handler.ssl.SslHandler$SslEngineType$1.unwrap(SslHandler.java:216)
        at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1297)
        at io.netty.handler.ssl.SslHandler.decodeNonJdkCompatible(SslHandler.java:1211)
        at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1245)
        at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:502)
        at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:441)
        ... 14 common frames omitted
2020-04-14 16:33:29,158 341327918 [CacheManager RUNNING] INFO  i.p.s.server.CacheManager - CacheManager: Gen: 309331-309331; Clients: 334 (309330-301212); Cache: Stored = 10291, Used = 438272, Reserved = 385024, Allocated = 197132288, Max = 11811160064.
2020-04-14 16:33:30,160 341328920 [CacheManager RUNNING] INFO  i.p.s.server.CacheManager - CacheManager: Gen: 309332-309332; Clients: 334 (309331-301212); Cache: Stored = 6579, Used = 434176, Reserved = 385024, Allocated = 197132288, Max = 11811160064.
2020-04-14 16:33:48,371 341347131 [epollEventLoopGroup-10-1] INFO  i.n.handler.logging.LoggingHandler - [id: 0x589ba91a, L:/0.0.0.0:12345] READ: [id: 0xd11c03a6, L:/172.17.0.38:12345 - R:/172.17.0.1:61471]



I'm using Self-Signed Certificates generated with openssl command :
`openssl req -nodes -new -x509 -keyout /tmp/certs/tls.key -days 365 -out /tmp/certs/ca-certificates/tls.crt -subj "/C=CA/ST=Quebec/L=Montreal/O=DellEMC/OU=Nautilus/CN=${NAUTILUS_DNS_NAME}"`

and the self-signed certificated is then loaded in the JVM during Docker Image Build
`RUN keytool -import -trustcacerts -noprompt -file /home/gradle/certs/tls.crt -alias nautilus_cert \
-storepass changeit -keystore /usr/local/openjdk-8/lib/security/cacerts`

--- a/README.md
+++ b/README.md
@@ -245,7 +245,12 @@ JAVA_TOOL_OPTIONS="${JAVA_TOOL_OPTIONS} -Dpravega.client.auth.token=YWRtaW46MTEx
 export JAVA_TOOL_OPTIONS
 ```
 
-Moreover, TLS host name verification is turned off by default to make it easier to run this tool. To enable hostname
+Note:
+* The token used in the example above is for `Basic` authentication, which is handled by the Password Auth
+Handler. The token's value equals Base64 encoded value of string `{username}:{password}`. The value shown in the example
+above is Base64 encoded value of string `admin:1111_aaaa`. If you are using another authentication method supported
+by a custom Pravega Auth Handler, generate a corresponding token, and specify it instead.
+* Moreover, TLS host name verification is turned off by default to make it easier to run this tool. To enable hostname
 verification, specify `-validateCertHostName true` option when executing it.
 
 ## Running in Docker

--- a/README.md
+++ b/README.md
@@ -246,11 +246,12 @@ export JAVA_TOOL_OPTIONS
 ```
 
 Note:
-* The token used in the example above is for `Basic` authentication, which is handled by the Password Auth
-Handler. The token's value equals Base64 encoded value of string `{username}:{password}`. The value shown in the example
-above is Base64 encoded value of string `admin:1111_aaaa`. If you are using another authentication method supported
-by a custom Pravega Auth Handler, generate a corresponding token, and specify it instead.
-* Moreover, TLS host name verification is turned off by default to make it easier to run this tool. To enable hostname
+* The specified token `YWRtaW46MTExMV9hYWFh` shown in the example above equals Base64 encoded value of credentials
+in `{username}:{password}` format supported by the specified authentication method `Basic`. The specified token is
+Base64 encoded value of string `admin:1111_aaaa`, represent the admin account available in the Password Auth Handler
+database. If you are using another authentication method supported by a custom Pravega Auth Handler, generate a
+corresponding token, and specify the token and method here instead.
+* TLS host name verification is turned off by default to make it easier to run this tool. To enable hostname
 verification, specify `-validateCertHostName true` option when executing it.
 
 ## Running in Docker

--- a/README.md
+++ b/README.md
@@ -238,7 +238,7 @@ for the client via the `JAVA_TOOL_OPTIONS` environment variable, before executin
 
 Below is an example:
 
-```
+```bash
 JAVA_TOOL_OPTIONS="-Djavax.net.ssl.trustStore=/path/to/client.truststore.jks"
 JAVA_TOOL_OPTIONS="${JAVA_TOOL_OPTIONS} -Dpravega.client.auth.method=Basic"
 JAVA_TOOL_OPTIONS="${JAVA_TOOL_OPTIONS} -Dpravega.client.auth.token=YWRtaW46MTExMV9hYWFh"

--- a/README.md
+++ b/README.md
@@ -92,6 +92,8 @@ usage: pravega-benchmark
                                      If >= 0 and using transactions,
                                      watermarks will be written on each
                                      commit.
+ -validateCertHostName               Whether to turn on host name verification
+                                     for TLS certificates
 ```
 ## Running Performance benchmarking
 
@@ -228,6 +230,23 @@ The `-throughput -1` specifies the writes tries to write the events at the maxim
 ### Recording the latencies to CSV files
 User can use the options `-writecsv  <file name>` to record the latencies of writers and `-readcsv <file name>` for readers.
 in case of End to End latency mode, if the user can supply only `-readcsv` to get the end to end latency in to the csv file.
+
+## Configuring Security Parameters
+
+To run against Security Enabled Pravega, you will need to specify TLS and authentication/authorization parameters
+for the client via the `JAVA_TOOL_OPTIONS` environment variable, before executing this tool.
+
+Below is an example:
+
+```
+JAVA_TOOL_OPTIONS="-Djavax.net.ssl.trustStore=/path/to/client.truststore.jks"
+JAVA_TOOL_OPTIONS="${JAVA_TOOL_OPTIONS} -Dpravega.client.auth.method=Basic"
+JAVA_TOOL_OPTIONS="${JAVA_TOOL_OPTIONS} -Dpravega.client.auth.token=YWRtaW46MTExMV9hYWFh"
+export JAVA_TOOL_OPTIONS
+```
+
+Moreover, TLS host name verification is turned off by default to make it easier to run this tool. To enable hostname
+verification, specify `-validateCertHostName true` option when executing it.
 
 ## Running in Docker
 

--- a/src/main/java/io/pravega/perf/PravegaPerfTest.java
+++ b/src/main/java/io/pravega/perf/PravegaPerfTest.java
@@ -12,6 +12,7 @@ package io.pravega.perf;
 
 import io.pravega.client.ClientConfig;
 import io.pravega.client.EventStreamClientFactory;
+import io.pravega.client.netty.impl.ConnectionFactoryImpl;
 import io.pravega.client.stream.ReaderGroup;
 import io.pravega.client.stream.impl.ClientFactoryImpl;
 import io.pravega.client.stream.impl.ControllerImpl;
@@ -28,6 +29,8 @@ import org.slf4j.LoggerFactory;
 import java.io.IOException;
 import java.net.URI;
 import java.net.URISyntaxException;
+import java.nio.file.Files;
+import java.nio.file.Paths;
 import java.util.List;
 import java.util.concurrent.Callable;
 import java.util.concurrent.ExecutionException;
@@ -98,6 +101,8 @@ public class PravegaPerfTest {
                 "If >0, watermarks will be read with a period of this many milliseconds.");
         options.addOption("reportingIntervalMillis", true, "period (in milliseconds) in which performance will be reported");
         options.addOption("createScope", true, "attempt to create Pravega scope(true by default)");
+        options.addOption("trustStoreLocation", true, "Location of the truststore to use when making TLS connections to server");
+        options.addOption("validateCertHostName", true, "Whether to turn on host name verification for TLS certificates");
 
         options.addOption("help", false, "Help message");
 
@@ -222,6 +227,8 @@ public class PravegaPerfTest {
         final long readWatermarkPeriodMillis;
         final int reportingInterval;
         final boolean createScope;
+        final String trustStoreLocation;
+        final boolean validateHostName;
 
         Test(long startTime, CommandLine commandline) throws IllegalArgumentException {
             this.startTime = startTime;
@@ -297,6 +304,14 @@ public class PravegaPerfTest {
             readWatermarkPeriodMillis = Long.parseLong(commandline.getOptionValue("readWatermarkPeriodMillis", "-1"));
 
             createScope = Boolean.parseBoolean(commandline.getOptionValue("createScope", "true"));
+
+            if (commandline.hasOption("trustStoreLocation")) {
+                trustStoreLocation = commandline.getOptionValue("trustStoreLocation");
+                log.debug("trustStoreLocation=" + trustStoreLocation);
+            } else {
+                trustStoreLocation = null;
+            }
+            validateHostName = Boolean.parseBoolean(commandline.getOptionValue("validateCertHostName", "false"));
 
             if (controllerUri == null) {
                 throw new IllegalArgumentException("Error: Must specify Controller IP address");
@@ -408,7 +423,6 @@ public class PravegaPerfTest {
                 return defaultValue;
             }
         }
-
     }
 
     static private class PravegaTest extends Test {
@@ -422,10 +436,19 @@ public class PravegaPerfTest {
             log.info("Test Parameters: {}", toString());
 
             final ScheduledExecutorService bgExecutor = Executors.newScheduledThreadPool(10);
+            ClientConfig.ClientConfigBuilder clientConfigBuilder = ClientConfig.builder().controllerURI(new URI(controllerUri));
+            if (controllerUri.startsWith("tls:") || controllerUri.startsWith("pravegas:")) {
+                if (trustStoreLocation != null) {
+                    if (!Files.exists(Paths.get(trustStoreLocation))) {
+                        throw new RuntimeException("Specified truststore file does not exist");
+                    }
+                    clientConfigBuilder.trustStore(trustStoreLocation);
+                }
+                clientConfigBuilder.validateHostName(false);
+            }
+            ClientConfig clientConfig = clientConfigBuilder.build();
             final ControllerImpl controller = new ControllerImpl(ControllerImplConfig.builder()
-                    .clientConfig(ClientConfig.builder().controllerURI(new URI(controllerUri)).build())
-                    .maxBackoffMillis(5000).build(),
-                    bgExecutor);
+                    .clientConfig(clientConfig).maxBackoffMillis(5000).build(), bgExecutor);
 
             streamHandle = new PravegaStreamHandler(scopeName, streamName, rdGrpName, controllerUri, segmentCount,
                     segmentScaleKBps, segmentScaleEventsPerSecond, scaleFactor, TIMEOUT, controller, bgExecutor, createScope);
@@ -439,12 +462,11 @@ public class PravegaPerfTest {
             }
 
             if (consumerCount > 0) {
-                readerGroup = streamHandle.createReaderGroup(!writeAndRead);
+                readerGroup = streamHandle.createReaderGroup(!writeAndRead, clientConfig);
             } else {
                 readerGroup = null;
             }
-
-            factory = new ClientFactoryImpl(scopeName, controller);
+            factory = new ClientFactoryImpl(scopeName, controller, new ConnectionFactoryImpl(clientConfig));
         }
 
         public List<WriterWorker> getProducers() {

--- a/src/main/java/io/pravega/perf/PravegaPerfTest.java
+++ b/src/main/java/io/pravega/perf/PravegaPerfTest.java
@@ -101,7 +101,6 @@ public class PravegaPerfTest {
                 "If >0, watermarks will be read with a period of this many milliseconds.");
         options.addOption("reportingIntervalMillis", true, "period (in milliseconds) in which performance will be reported");
         options.addOption("createScope", true, "attempt to create Pravega scope(true by default)");
-        options.addOption("trustStoreLocation", true, "Location of the truststore to use when making TLS connections to server");
         options.addOption("validateCertHostName", true, "Whether to turn on host name verification for TLS certificates");
 
         options.addOption("help", false, "Help message");
@@ -227,7 +226,6 @@ public class PravegaPerfTest {
         final long readWatermarkPeriodMillis;
         final int reportingInterval;
         final boolean createScope;
-        final String trustStoreLocation;
         final boolean validateHostName;
 
         Test(long startTime, CommandLine commandline) throws IllegalArgumentException {
@@ -304,13 +302,6 @@ public class PravegaPerfTest {
             readWatermarkPeriodMillis = Long.parseLong(commandline.getOptionValue("readWatermarkPeriodMillis", "-1"));
 
             createScope = Boolean.parseBoolean(commandline.getOptionValue("createScope", "true"));
-
-            if (commandline.hasOption("trustStoreLocation")) {
-                trustStoreLocation = commandline.getOptionValue("trustStoreLocation");
-                log.debug("trustStoreLocation=" + trustStoreLocation);
-            } else {
-                trustStoreLocation = null;
-            }
             validateHostName = Boolean.parseBoolean(commandline.getOptionValue("validateCertHostName", "false"));
 
             if (controllerUri == null) {
@@ -436,19 +427,12 @@ public class PravegaPerfTest {
             log.info("Test Parameters: {}", toString());
 
             final ScheduledExecutorService bgExecutor = Executors.newScheduledThreadPool(10);
-            ClientConfig.ClientConfigBuilder clientConfigBuilder = ClientConfig.builder().controllerURI(new URI(controllerUri));
-            if (controllerUri.startsWith("tls:") || controllerUri.startsWith("pravegas:")) {
-                if (trustStoreLocation != null) {
-                    if (!Files.exists(Paths.get(trustStoreLocation))) {
-                        throw new RuntimeException("Specified truststore file does not exist");
-                    }
-                    clientConfigBuilder.trustStore(trustStoreLocation);
-                }
-                clientConfigBuilder.validateHostName(false);
-            }
-            ClientConfig clientConfig = clientConfigBuilder.build();
+            ClientConfig clientConfig = ClientConfig.builder().controllerURI(new URI(controllerUri))
+                    .validateHostName(validateHostName).build();
             final ControllerImpl controller = new ControllerImpl(ControllerImplConfig.builder()
-                    .clientConfig(clientConfig).maxBackoffMillis(5000).build(), bgExecutor);
+                    .clientConfig(clientConfig)
+                    .maxBackoffMillis(5000).build(),
+                    bgExecutor);
 
             streamHandle = new PravegaStreamHandler(scopeName, streamName, rdGrpName, controllerUri, segmentCount,
                     segmentScaleKBps, segmentScaleEventsPerSecond, scaleFactor, TIMEOUT, controller, bgExecutor, createScope);

--- a/src/main/java/io/pravega/perf/PravegaStreamHandler.java
+++ b/src/main/java/io/pravega/perf/PravegaStreamHandler.java
@@ -148,10 +148,9 @@ public class PravegaStreamHandler {
         }
     }
 
-    ReaderGroup createReaderGroup(boolean reset) throws URISyntaxException {
+    ReaderGroup createReaderGroup(boolean reset, ClientConfig clientConfig) throws URISyntaxException {
         if (readerGroupManager == null) {
-            readerGroupManager = ReaderGroupManager.withScope(scope,
-                    ClientConfig.builder().controllerURI(new URI(controllerUri)).build());
+            readerGroupManager = ReaderGroupManager.withScope(scope, clientConfig);
             rdGrpConfig = ReaderGroupConfig.builder()
                     .stream(Stream.of(scope, stream)).build();
         }

